### ISSUE TITLE
links posted in Journal are not clickable and the link created goes to a page on the site

### ISSUE THREAD

**Issue by thabaum** [dnnsoftware/Dnn.Platform#3308]
## Description of bug
Posting a link to a site in the journal posts content about the site, however the link to the site is not correct when you click on it, and the link in journal does not work.

## Steps to reproduce
List the steps to reproduce the behavior:
1. Go to journal
2. post a link
3. click on link
4. takes you to a page on your site 

## Current result
Link does not work properly that is created, and the link posted is not clickable

## Expected result
A possibly shortened clickable link with information about the page and a link to the page posted in the header

## Screenshots
If applicable, provide screenshots to help explain your problem.
![image](https://user-images.githubusercontent.com/13577556/69024045-f1c0f600-0975-11ea-99a2-fbe5dc139050.png)

**Comment by valadas** [comment]
I can reproduce in 9.4.2

**Comment by thabaum** [comment]
The images are not working as well.  I may have found a solution but I need to figure out how to test faster so I can see if it indeed is a solution in the short days we live in...  Both links are stripped for image source and URL link.  I also have another issue relating to this area I will try to knock em both out while looking here.  I want to fix an issue for the DNN Events calendar.  https://github.com/DNNCommunity/DNN.Events/issues/40

This one has been an issue forever, i feel like just stripping the HTML or figuring out why this is different from other posts.   So I am trying to work on both of these concerns now along with a few others for the events modules hoping I can help out on it.  DNN Events needs some love too :)

**Comment by thabaum** [comment]
An update here so this does not go stale:

The journal is not saving the URL: in the database.  I have read the module source code a few times and believe that I am getting warmer as to where this event happens.  I think I need a better testing environment setup to really tackle things like this one.

Some odd behavior when you manually set the URL: part that is missing in the databse journal item data that I noticed:

 If you have a url like www.therestoftheurl.com instead of one with http/https in front of it, the journal tries to load a page from within the site www.dnnsite.com/www.therestoftheurl.com but if you have http:// or https:// added it will link to domains off the website.  I would prefer not to have either http or https and load a new browser tab with just the domain and let the web server that handles the domain put http or https on the url as need if possible?!

Maybe this might help someone come up with a solution as well.  Thank you for any help and support here.

**Comment by david-poindexter** [comment]
@thabaum I have not dug into the code myself, but a guiding principle across DNN (and most other web apps) is that when storing URL field data, if it is not prefixed by `https://`, `http://` or `//`, then it is not an **absolute** URL.  It is instead assumed to be a **relative** URL.  Therefore, the behavior being reported is actually what I would expect.  

As it relates to the URL's "preview image" that is usually provided via a combination of page data and metadata:

* Title
* Description
* og:title
* og:description
* og:url
* og:image
* favicon

Again, I am not sure what this code is reliant on in DNN, but it is most likely one or more of the above.  The **og:** prefixed metadata is based on OpenGraph.

That said, is this really a bug?

**Comment by valadas** [comment]
The protocol should be part of an url, an url is in fact not even for web, it could even be ftp://somedomain.com/somefolder so yeah, the protocol is definitely a needed part of any url.

**Comment by david-poindexter** [comment]
@valadas ultimately yes - of course.  😄 

I'm just saying that DNN most likely supports relative and absolute URLs.  Whether or not it should in this context is probably where the focus needs to be.  I would think in this context it should only accommodate/support absolute URLs.

**Comment by thabaum** [comment]
Posting a "link" in the journal works when you just type "www.domain.com" and so DNN journal does not know HTTP or HTTPS but still knows it is a link and then it strips things // and prior to the www I believe I seen in the code.  There are a few areas that are of interest I will make references here soon maybe it can help shine light here. 

DNN Journal would have to add an HTTP:// or HTTPS:// without knowing which to put I am guessing the HTTP:// would be default.  I was trying to find where it adds this, i found some code that strips it and adds again.  When it grabs images and description of the page it has the URL in memory so need to know why that URL is not getting set in the Database.

**Comment by thabaum** [comment]
@david-poindexter 

The bug here is the link does not work, it is blank when you post a link in a journal post.  Image and description of the website works but if you click on the link created for the journal post it goes no where.

To reproduce post a link in the journal and try to click on the url that is created for a link to the website that is entered.  The href="" which in the database it is not being set.   I can manually add the link that is missing in the database and everything works for that post.  This is when I found it the link must have HTTP:// or HTTPS:// to make the link work in the journal because when I set it to just have a domain name without those it would go to a local page on the website creating a page not found error.  

This is just a clue to what needs to happen prior to setting the URL in the database is there needs to be HTTP:// added to the front of the website link for DNN Journal to allow it to work AND it is not getting set at all either way so there are two things going on here that must happen for it to work right.

1.  the www.domain.com link needs be http://www.domain.com to work
2.  the domain after it has been properly created in a variable needs to add that to the database when created the post inside the journal module on a dnn page.

Some reason the url is not being posted to the database in any way, it is blank so when you click on the link created in the journal it goes no where.

This is a bug and needs some love.

**Comment by thabaum** [comment]
After digging a bit further I believe I may have found the issue potentially with your help?  I need to know `!IsAllowedLink` part of the code below is something that is getting set somehow somewhere to where it is making the ItemData.Url string empty.

Is there a way to test this further quickly maybe by commenting those lines out below.  And if there is a proper way to deal with this if there is an issue here.  Thank you!

https://github.com/dnnsoftware/Dnn.Platform/blob/develop/DNN%20Platform/Modules/Journal/ServicesController.cs#L217-L221     

Line: 217
         ```
  if (!IsAllowedLink(ji.ItemData.Url))
                {
                    ji.ItemData.Url = string.Empty;
                }
```

**Comment by thabaum** [comment]
Keeping this from going stale, I am going to give it a go here soon pick up where I left off in the next couple weeks.   This, social groups and DNN events not stripping the HTML are issues related I will be hoping to address for some version 9 fixes.  Thank you.

**Comment by thabaum** [comment]
Issue looked a little confusing. I updated the issue along with adding some additional context with things I believe may be happening here to create the issue.  If you post a link in the journal the Url is not being saved to the data items in the database.  So then the url is blank when the journal creates post entry to be viewed.  If you add an address with Http:// or Https:// in the database manually the address appears and will work right.  

Possibility is the https:// and http;// get checked and if they have anything with a // it is removed.  But I don't think http:// or https:// is ever added back again to the link data and then when it needs to get saved it gets checked to see if it has // but it has been removed when it was cleaned of it therefore a url is never stored it the database.   A www does not have this so it fails as well getting saved.  If I remember right this is what I believe is the current issue best explained I can.

I am going to hopefully figure a way to step through this again and understand what is missing here.  I will be giving this all another go here in the next week.

**Comment by thabaum** [comment]
OK just so this is not stale... my situation is getting a better environment to test for a solution for this issue.  I only have so much time and resources to spend to make something happen and I have done it a few times in the past but it is a process to setup and get things running on my machine which is what has been dragging me behind here. I am a past unix, C and pascal programmer give me a brake :)... I do need to get this completed I need this to function properly.  I did put some notes above talking to myself... I will review after I get down editing this part of the platform, building and debugging it.  Part I have the issue with is getting the site going.

What I believe is a solution here... is steps, videos no longer than 3 minutes long.  Each is a step towards setting up the environment.  Everyone may have an issue with one step but you have to watch 5 20 minute to 1 hour videos and still may not get the answer to making it happen.  Anyone have a link to a good relevant version of setting up DNN so you can debug the platform in a productive way?  I just need an example of making a change to a file saving it and running a site to view the change.  I am guessing otherwise and I think whatever I am missing everyone else takes for granted ;)

**Comment by valadas** [comment]
Did you read the build instructions here, there are explanations for creating a full release, updating an existing development site or building for debug. Following the build instructions for debug is probably what you want and then you would put a breakpoint and attach to the w3wp process for the site and it should work. If you have problems, please let us know.

**Comment by thabaum** [comment]
Thank you @valadas it has been a real challenge, seems like at one time I had it all right but then just blanked on how to do it again with so much time between attempts not doings this daily...

I will get back with any questions I have along the way.  It would be fun to get into debugging this in a pro-efficient way... and I think I just need to practice doing it for a while a lot.  I am getting warmer to VS and GitHub so that helps not having that frustration on top of this bug :)

**Comment by thabaum** [comment]
https://github.com/dnnsoftware/Dnn.Platform/blob/develop/.github/BUILD.md

That is the link to get environment setup,

I found my problem, I never knew what Cake was... I always wondered and figured it had to do with things that it did but nothing about actually configuring it for my environment.

This unpacks the web files so I can see changes, this was my missing link.

```
missingLink == Cake;
```

**Comment by thabaum** [comment]
I am going to do a recap here soon as I set up hoping this week get somewhere on this issue.

**Comment by thabaum** [comment]
OK, its cold again... I am going to give this a go!

**Comment by thabaum** [comment]
[Line 422 ](https://github.com/dnnsoftware/Dnn.Platform/blob/7f8df1689563483a4005c1b686a8ab09ae47f3ca/DNN%20Platform/Modules/Journal/ServicesController.cs#L422)

```
                if (!IsAllowedLink(ji.ItemData.Url))
                {
                    //ji.ItemData.Url = string.Empty;
                }

```

I commented out the entire if statement and the URL information was not submitted to the database.
I commented out just the line inside the if statement as shown above and it works!

What should we do here to make this right.  I found a solution, delete this one line.  But the if statement has to stay here or the url is never submitted to the database array information for `ItemData`

Make Sense?  I want to submit a PR but I dont think I have this 100% yet.   Any input on what should be done here to make it a success!

**Comment by thabaum** [comment]
![image](https://user-images.githubusercontent.com/13577556/145157091-81afd110-cc69-481a-b969-deb574c171e7.png)

**Comment by thabaum** [comment]
[Line 318](https://github.com/dnnsoftware/Dnn.Platform/blob/7f8df1689563483a4005c1b686a8ab09ae47f3ca/DNN%20Platform/Modules/Journal/ServicesController.cs#L318)
```
        private static bool IsAllowedLink(string url)
        {
            return !string.IsNullOrEmpty(url) && !url.Contains("//");
        }

```

If you are posting a Url, Of coarse it will have "//" in an address like http://domain.com.

Another issue is it puts HTTP instead of HTTPS, however most sites forward to HTTPS I entered HTTPS site, What you enter is what should be entered into the database.

**Comment by thabaum** [comment]
what is funny... github makes a link out of a `www` in places, I feel DNN was going to be a little better having `www.`

**Comment by thabaum** [comment]
I double tested this solution and seems to resolve things just fine.  I like how it helps the ease of extending future features with journal input as well.  Thank you all for reviewing this.

**Comment by dustinp26** [comment]
For what its worth, you can use this javascript/jquery to make the links clickable. More a patch however, If the users was to delete the actual link text before posting it won't work.

		var targetDiv = document.querySelectorAll("[id^='jid-'] .journalitem p:first-child");
		for(var i=0; i<targetDiv.length;i++){
			
			var sumbittedUrls = $(targetDiv[i]).text();
		  	var urls = sumbittedUrls.match(/(ftp|http|https):\/\/(\w+:{0,1}\w*@)?(\S+)(:[0-9]+)?(\/|\/([\w#!:.?+=&%@!\-\/]))?/);

			if($(targetDiv[i]).parent().find('.jlink a').length !== 0) {
		  		$(targetDiv[i]).parent().find('.jlink a').attr('href', urls[0]);	  

				$('.journalitem p:contains('+(urls[0])+')').each(function(){
					$(this).html($(this).html().split(urls[0]).join(""));
				});
				
			}

		}
